File: code/MCMC.py
import numpy as np
from scipy.stats import gamma 


class MCMC:
    """
    Class for MCMC sampling
    """
    def __init__(
        self, 
        model, 
        data, 
        qpriors, 
        qstart, 
        nsamples=100, 
        lstm_model={}, 
        adapt_interval=10, 
        verbose=True):

        self.model          = model
        self.qstart         = qstart
        self.qpriors        = qpriors
        self.nsamples       = nsamples # number of samples to take during MCMC algorithm
        self.nburn          = int(nsamples/2) # number of samples to discard during burn-in period
        self.verbose        = verbose
        self.adapt_interval = adapt_interval
        self.data           = data
        self.lstm_model     = lstm_model
        self.n0             = 0.01

        return

    def sample(self):
        """ 
        The code provided seems to be part of a sampling algorithm 
        that uses the Metropolis-Hastings algorithm to generate samples from a distribution. 
        Markov Chain Monte Carlo (MCMC) method for parameter estimation.
        """
        # Evaluate the model with the original dc value
        if self.lstm_model:
            acc_ = self.model.rom_evaluate(self.lstm_model)[1]
        else:
            acc_ = self.model.evaluate()[1]

        # Perturb the dc value
        self.model.Dc *= (1 + 1e-6)

        # Evaluate the model with the perturbed dc value
        if self.lstm_model:
            acc_dq_ = self.model.rom_evaluate(self.lstm_model)[1]
        else:
            acc_dq_ = self.model.evaluate()[1]

        # Extract the values and reshape them to a 1D array
        acc = acc_.reshape(1, -1)
        acc_dq = acc_dq_.reshape(1, -1)

        # Compute the variance of the noise
        self.std2 = [np.sum((acc - self.data) ** 2, axis=1).item()/(acc.shape[1] - len(self.qpriors))]

        # Compute the covariance matrix
        X                  = ((acc_dq - acc) / (self.model.Dc * 1e-6)).T
        X                  = np.linalg.inv(np.dot(X.T, X))
        self.Vstart        = self.std2[-1] * X 
        qstart_vect        = np.array([[self.qstart]])
        self.qstart_limits = np.array([[self.qpriors[1], self.qpriors[2]]])
        qparams            = np.copy(qstart_vect) # Array of sampled parameters
        Vold               = np.copy(self.Vstart) # Covariance matrix of previously sampled parameters
        Vnew               = np.copy(self.Vstart) # Covariance matrix of current sampled parameters
        SSqprev            = self.SSqcalc(qparams) # Squared error of previously sampled parameters
        iaccept            = 0 # Counter for accepted samples

        for isample in np.arange(self.nsamples):
            # Sample new parameters from a normal distribution with mean being the last element of qparams
            q_new = np.reshape(np.random.multivariate_normal(qparams[:,-1],Vold),(-1,1)) 

            # Accept or reject the new sample based on the Metropolis-Hastings acceptance rule
            accept,SSqnew = self.acceptreject(q_new,SSqprev,self.std2[-1])

            # Print some diagnostic information
            print(isample,accept)
            print("Generated sample ---- ",np.asscalar(q_new))

            # If the new sample is accepted, 
            # add it to the list of sampled parameters
            if accept:
                qparams    = np.concatenate((qparams,q_new),axis=1)
                SSqprev    = SSqnew.copy()
                iaccept    += 1
            else:
                # If the new sample is rejected, 
                # add the previous sample to the list of sampled parameters
                q_new      = np.reshape(qparams[:,-1],(-1,1))
                qparams    = np.concatenate((qparams,q_new),axis=1)

            # Update the estimate of the standard deviation
            aval           = 0.5*(self.n0+len(self.data))
            bval           = 0.5*(self.n0*self.std2[-1]+SSqprev)
            self.std2.append(1/gamma.rvs(aval,scale=1/bval,size=1)[0])

            # Update the covariance matrix if it is time to adapt it
            if (isample+1) % self.adapt_interval == 0:
                try:
                    Vnew = 2.38**2/len(self.qpriors.keys())*np.cov(qparams[:,-self.adapt_interval:])
                    if qparams.shape[0]==1:
                        Vnew = np.reshape(Vnew,(-1,1))
                    Vnew = np.linalg.cholesky(Vnew)
                    Vold = Vnew.copy()
                except:
                    pass

        # Print acceptance ratio
        print("acceptance ratio:",iaccept/self.nsamples)

       # Trim the estimate of the standard deviation to exclude burn-in samples  
        self.std2 = np.asarray(self.std2)[self.nburn:] 
        
        # Return accepted samples
        return qparams[:,self.nburn:]

    def acceptreject(self, q_new, SSqprev, std2):
        """ 
        The acceptreject function checks if a proposed sample falls within the limits 
        and decides whether to accept or reject it 
        based on the acceptance probability and a random number. 
        It returns a boolean value indicating acceptance 
        and the sum of squares error of the accepted or previous proposal.
        """
        # Check if the proposal values are within the limits
        accept = np.all((q_new > self.qstart_limits[:, 0]) & (q_new < self.qstart_limits[:, 1]), axis=0)

        if accept:
            # Compute the sum of squares error of the new proposal
            SSqnew = self.SSqcalc(q_new)

            # Compute the acceptance probability
            accept_prob = np.clip(0.5 * (SSqprev - SSqnew) / std2, -np.inf, 0)

            # Check if the proposal is accepted 
            # based on the acceptance probability and a random number
            accept = accept_prob > np.log(np.random.rand(1))[0]

        if accept:
            # If accepted, return the boolean True and the sum of squares error of the new proposal
            return accept, SSqnew
        else:
            # If rejected, return the boolean False and the sum of squares error of the previous proposal
            return accept, SSqprev

    def SSqcalc(self, q_new):
        """ 
        The SSqcalc function updates the Dc parameter of the model with a proposed value, 
        evaluates the model's performance, and computes the sum of squares error 
        between the model's accuracy and the data.
        """
        # Update the Dc parameter of the model with the new proposal
        self.model.Dc = q_new[0,]

        # Evaluate the model's performance on the problem type and LSTM model
        if self.lstm_model:
            acc = self.model.rom_evaluate(self.lstm_model)[1]
        else:
            acc = self.model.evaluate()[1]

        # Compute the sum of squares error between the model's accuracy and the data
        SSq = np.sum((acc.reshape(1, -1) - self.data)**2, axis=1, keepdims=True)

        return SSq

File: code/MCMC.cpp
#include <iostream>
#include <random>
#include <cmath>
#include <vector>
#include </opt/homebrew/Cellar/eigen/3.4.0_1/include/eigen3/Eigen/Dense>
#include </opt/homebrew/Cellar/eigen/3.4.0_1/include/eigen3/Eigen/Cholesky>
#include </opt/homebrew/Cellar/eigen/3.4.0_1/include/eigen3/unsupported/Eigen/MatrixFunctions>

using namespace std;

class MCMC {
private:
    // Define class variables
    RateStateModel model;
    double qstart;
    vector<double> qpriors;
    int nsamples;
    int nburn;
    bool verbose;
    int adapt_interval;
    vector<double> data;
    double n0;
    vector<double> std2;
    Eigen::MatrixXd Vstart;
    Eigen::MatrixXd qstart_limits;

public:
    MCMC(RateStateModel model, vector<double> data, vector<double> qpriors, double qstart,
         int nsamples = 100, int adapt_interval = 10, bool verbose = true)
        : model(model), data(data), qpriors(qpriors), qstart(qstart), nsamples(nsamples),
          nburn(nsamples / 2), verbose(verbose), adapt_interval(adapt_interval),
          n0(0.01) {}

    Eigen::MatrixXd sample() {
        // Evaluate the model with the original dc value
        Eigen::MatrixXd acc_ = model.evaluate()[1];

        // Perturb the dc value
        model.Dc *= (1 + 1e-6);

        // Evaluate the model with the perturbed dc value
        Eigen::MatrixXd acc_dq_ = model.evaluate()[1];

        // Extract the values and reshape them to a 1D array
        Eigen::MatrixXd acc = acc_.reshape(1, -1);
        Eigen::MatrixXd acc_dq = acc_dq_.reshape(1, -1);

        // Compute the variance of the noise
        std2.push_back((acc - data).array().square().sum() / (acc.cols() - qpriors.size()));

        // Compute the covariance matrix
        Eigen::MatrixXd X = ((acc_dq - acc) / (model.Dc * 1e-6)).transpose();
        Eigen::MatrixXd XTX = X.transpose() * X;
        Vstart = std2.back() * XTX.inverse();
        Eigen::MatrixXd qstart_vect = Eigen::MatrixXd::Constant(1, 1, qstart);
        qstart_limits = Eigen::MatrixXd::Constant(1, 2, qpriors[0], qpriors[1]);
        Eigen::MatrixXd qparams(qstart_vect.rows(), nsamples);
        qparams.col(0) = qstart_vect;
        Eigen::MatrixXd Vold = Vstart;
        Eigen::MatrixXd Vnew = Vstart;
        double SSqprev = SSqcalc(qparams.col(0))(0);
        int iaccept = 0;

        for (int isample = 1; isample < nsamples; isample++) {
            // Sample new parameters from a normal distribution with mean being the last element of qparams
            Eigen::MatrixXd q_new = qparams.col(isample - 1) +
                                    Eigen::MatrixXd::NullaryExpr(qparams.rows(), 1,
                                    [&]() { return normal_distribution<>(0.0, 1.0)(random_engine); }) *
                                    Vold.llt().matrixL();

            // Accept or reject the new sample based on the Metropolis-Hastings acceptance rule
            double SSqnew;
            bool accept;
            tie(accept, SSqnew) = acceptreject(q_new, SSqprev, std2.back());

            // If the new sample is accepted, add it to the list of sampled parameters
            if (accept) {
                qparams.col(isample) = q_new;
                SSqprev = SSqnew;
                iaccept++;
            } else {
                // If the new sample is rejected, add the previous sample to the list of sampled parameters
                qparams.col(isample) = qparams.col(isample - 1);
            }

            // Update the estimate of the standard deviation
            double aval = 0.5 * (n0 + data.size());
            double bval = 0.5 * (n0 * std2.back() + SSqprev);
            std2.push_back(1.0 / gamma_distribution<>(aval, 1.0 / bval)(random_engine));

            // Update the covariance matrix if it is time to adapt it
            if ((isample + 1) % adapt_interval == 0) {
                try {
                    Vnew = 2.38 * 2 / qpriors.size() * qparams.rightCols(adapt_interval).cov();
                    if (qparams.rows() == 1) {
                        Vnew.resize(1, 1);
                    }
                    Vnew = Vnew.llt().matrixL();
                    Vold = Vnew;
                } catch (...) {
                    // Ignore any errors
                }
            }
        }

        // Print acceptance ratio
        cout << "Acceptance ratio: " << static_cast<double>(iaccept) / nsamples << endl;

        // Trim the estimate of the standard deviation to exclude burn-in samples
        std2.erase(std2.begin(), std2.begin() + nburn);

        // Return accepted samples
        return qparams.rightCols(nsamples - nburn);
    }

private:
    mt19937 random_engine{random_device{}()};

    tuple<bool, double> acceptreject(const Eigen::MatrixXd& q_new, double SSqprev, double std2) {
        // Check if the proposal values are within the limits
        bool accept = (q_new.array() > qstart_limits(0, 0)) && (q_new.array() < qstart_limits(0, 1));

        if (accept) {
            // Compute the sum of squares error of the new proposal
            double SSqnew = SSqcalc(q_new)(0);

            // Compute the acceptance probability
            double accept_prob = min(0.5 * (SSqprev - SSqnew) / std2, 0.0);

            // Check if the proposal is accepted based on the acceptance probability and a random number
            accept = accept_prob > log(uniform_real_distribution<>(0.0, 1.0)(random_engine));
        }

        if (accept) {
            // If accepted, return the boolean true and the sum of squares error of the new proposal
            return make_tuple(true, SSqnew);
        } else {
            // If rejected, return the boolean false and the sum of squares error of the previous proposal
            return make_tuple(false, SSqprev);
        }
    }

    Eigen::VectorXd SSqcalc(const Eigen::MatrixXd& q_new) {
        // Update the Dc parameter of the model with the new proposal
        model.Dc = q_new(0, 0);

        // Evaluate the model's performance on the problem
        Eigen::MatrixXd acc = model.evaluate()[1];

        // Compute the sum of squares error between the model's accuracy and the data
        return (acc - data).array().square().colwise().sum();
    }
};